### PR TITLE
[REVIEW] [FEA] Symmetrize a COO

### LINKED ISSUE #310: [FEA] symmetrize feature
Also known as M = M+M^T op

Input : COO 
Output: symmetric COO (ie. each undirected edge is stored as an edge in both direction)

This should work for :
- undirected graph input (provided as an edge list where `src[i] dst[i]` implicitly represents `dst[i] src[i]`)
- directed graph input that is not symmetric

From Nx perspective, that means combining a `digraph` with its `reverse_view`

We need a cuda implementation as well as Python bindings.

### LINKED ISSUE #430: [FEA] Symmetrize edge list
For each edge (u, v, data) on input, there should be two directed edges (u, v, data) and (v, u, data) on output. 

- The input edge list could be directed or undirected 
- There should not be duplicates in input or output. 
- There might be self-loop in input or output. 

I think the API should be similar to the renumbering function ie. `src_s, dst_s = symmetrize(sources, destinations)` 

This could be implemented by adding (v, u) for each (u, v) in the edge list and then calling `cudf.drop_duplicates`.

This task also involves doing the python bindings, tests and documentation.

See also this networkx related feature : 
https://networkx.github.io/documentation/latest/reference/classes/generated/networkx.Graph.to_directed.html

### PR DESCRIPTION
**Pull request by ChuckHastings**
Python implementation (using cudf calls) to symmetrize a coo.

cudf already provides all of the necessary functions to perform this.  It seems unlikely - due to the simplicity of the task - that a custom implementation would be significantly more efficient - but if so that can be done as an optimization task later.

closes #430 
closes #310 

Note that this PR also adds a small directed graph to the python test data sets.  This graph could be used for testing directed analytics (e.g. SCC)

### CODE REVIEW

**Comment by ChuckHastings** [comment]
Note, I have disabled one of our usual python tests.  There's a note in the code with a reference to https://github.com/rapidsai/cudf/issues/2636 which is the cudf bug report for the problem.  Once that bug is resolved we should update this code to perform the usual test.  If we approve this PR I suggest we create cugraph issue linking to the cudf issue so that we remember to fix it later.

File: python/cugraph/structure/symmetrize.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+# limitations under the License.
+
+from cugraph.structure.graph import null_check
+import cudf
+
+
+def symmetrize_df(df, src_name, dst_name):
+    """
+    Take a COO stored in a dataframe, along with the column names of
+    the source and destination columns assumed to represent a directed graph
```
Comment (afender) [inline python/cugraph/structure/symmetrize.py]: Should we get rid of "assumed to represent a directed graph"? It seems to me that a user may see her input as **undirected** (ie. storing only one direction in the edge list) and we would properly return a new graph with directed edges appearing in both directions.
This also appears in `symmetrize` doc.

File: python/cugraph/tests/test_symmetrize.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+from librmm_cffi import librmm as rmm
+from librmm_cffi import librmm_config as rmm_cfg
+
+
+def test_version():
+    gc.collect()
+    cugraph.__version__
+
+
+DATASETS = ['../datasets/karate',
```
Comment (afender) [inline python/cugraph/tests/test_symmetrize.py]: While `karate.csv` is an undirected graph, the version we have in the repo already stores an undirected edge as a directed edge in both directions (notice the 78 undirected edges in the `.mtx` vs. 156 lines in the `.csv` file). So, the symmetrize feature should not do anything on this dataset.

**Comment by afender** [comment]
rerun tests

File: python/cugraph/structure/symmetrize.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        Name of the column in the data frame containing the dest ids
+
+    Examples
+    --------
+    >>> M = cudf.read_csv('datasets/karate.csv', delimiter=' ',
+    >>>                   dtype=['int32', 'int32', 'float32'], header=None)
+    >>> sym_df = cugraph.symmetrize(M, '0', '1')
+    >>> G = cugraph.Graph()
+    >>> G.add_edge_list(M['0]', M['1'], M['2'])
+    """
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply python/cugraph/structure/symmetrize.py]: Made this change.

File: python/cugraph/structure/symmetrize.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    Parameters
+    ----------
+    df : cdf.DataFrame
+        Input data frame containing COO.  Columns should contain source
+        ids, destination ids and any properties associated with the
+        edges.
+    src_name : string
+        Name of the column in the data frame containing the source ids
+    dst_name : string
+        Name of the column in the data frame containing the dest ids
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply python/cugraph/structure/symmetrize.py]: Done

File: python/cugraph/tests/test_symmetrize.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    assert(rmm.is_initialized())
+
+    cu_M = utils.read_csv_file(graph_file+'.csv')
+
+    sym_sources, sym_destinations = cugraph.symmetrize(cu_M['0'], cu_M['1'])
+
+    #
+    #  Check to see if all pairs in sources/destinations exist in
+    #  both directions
+    #
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply python/cugraph/tests/test_symmetrize.py]: Figured out a good way to do the join to compare weights.  Added a specific python test with different weights.

Good news is, the test that compares weights works... it found a bug in the original code!

File: python/cugraph/structure/symmetrize.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        if (name == src_name):
+            gdf[src_name] = df[src_name].append(df[dst_name],
+                                                ignore_index=True)
+        elif (name == dst_name):
+            gdf[dst_name] = df[dst_name].append(df[src_name],
+                                                ignore_index=True)
+        else:
+            gdf[name] = df[name].append(df[name], ignore_index=True)
+
+    return gdf.drop_duplicates(subset=[src_name, dst_name], keep='first')
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply python/cugraph/structure/symmetrize.py]: Used this example as the basis for a test.  You are correct, it generates the incorrect answer always in this test case.

Changed the implementation to take the minimum value so we will be deterministic.

File: python/cugraph/tests/test_symmetrize.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    #  we can join the data frames (using the DataFrame.merge function).
+    #  The symmetrize function should contain every edge that was contained
+    #  in the input data.  So if we join the input data with the output
+    #  the length of the data frames should be equal.
+    #
+    sym_df = cudf.DataFrame([('src', sym_src),
+                             ('dst', sym_dst),
+                             ('weight', sym_w)])
+
+    #  TODO:  How to test weights...
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply python/cugraph/tests/test_symmetrize.py]: Found a way to use the join to do this.

**Comment by ChuckHastings** [comment]
rerun tests

**Comment by raydouglass** [comment]
rerun tests

**Comment by afender** [comment]
rerun tests

**Pull request merged**